### PR TITLE
fork-choice interface

### PR DESCRIPTION
**Pull request by S1nus**
Fork-choice package added to the repo. It supports defining any fork-choice rule as a `func ([]*types.Block) (*types.Block, bool)`, it works similarly to the DA interface.

The PR also integrates the fork-choice interface into `BlockManager`.

This PR also includes an implementation of the `FirstOrderedPure` fork-choice rule, and a placeholder for the `FirstOrderedCentralized`, whose implementation is currently blocked by #735.

### CODE REVIEW

**Comment by gupadhyaya** [comment]
@S1nus i think you will have to use default FCR in all the tests that create newFullNode. otherwise, there won't be any FCR, hence leading to nil pointer error.

**Comment by S1nus** [comment]
Newest commit changes FCRs from a function signature to an interface type. Certain schemes will want to have state, hence why it makes sense to represent the FCR as a struct. There might also be other functions besides `Apply()`.

**Comment by S1nus** [comment]
another idea: the fork-choice rule configuration option may be better suited as a parameter in the genesis block, rather than Rollkit's config

File: block/manager.go [hunk 1]
```diff
@@ -134,6 +136,13 @@ func NewManager(
 		}
 	}
 
+	fcr, ok := fork_choice.GetRule(conf.ForkChoiceRule)
+	if !ok {
+		fcr, ok = fork_choice.GetRule(config.DefaultNodeConfig.BlockManagerConfig.ForkChoiceRule)
```
Comment (gupadhyaya) [inline block/manager.go:141]: multiple calls to `GetRule` is not needed. when `!ok` simply assign `firstOrderedPure` to `fcr`?

File: block/manager.go [hunk 2]
```diff
@@ -291,7 +301,10 @@ func (m *Manager) SyncLoop(ctx context.Context, cancel context.CancelFunc) {
 				"daHeight", daHeight,
 				"hash", block.Hash(),
 			)
-			m.syncCache[block.SignedHeader.Header.BaseHeader.Height] = block
+			if m.syncCache[block.SignedHeader.Header.BaseHeader.Height] == nil {
```
Comment (gupadhyaya) [inline block/manager.go:304]: `height := block.SignedHeader.Header.BaseHeader.Height` and use `height` in multiple places

File: block/manager.go [hunk 3]
```diff
@@ -339,11 +352,11 @@ func (m *Manager) trySyncNextBlock(ctx context.Context, daHeight uint64) error {
 	var commit *types.Commit
 	currentHeight := m.store.Height() // TODO(tzdybal): maybe store a copy in memory
 
-	b1, ok1 := m.syncCache[currentHeight+1]
+	b1, ok1 := m.fcr.Apply(m.syncCache[currentHeight+1])
```
Comment (gupadhyaya) [inline block/manager.go:355]: sync to the latest main, as this logic will change a bit: https://github.com/rollkit/rollkit/blob/main/block/manager.go#L313

**Pull request closed without merging**